[158] Refresh compiled rules if stale.

AshishPrasad opened the pull request:
Solves #158.

### Changes

1. Attach timestamp in `_workflowRules` and `_compileRules` in `RulesCache.cs` to determine if the compile rule is stale.
2. Fix the bug for not refreshing compiled rules if workflow rules have been added.
3. Introduce a new method `AddOrUpdateWorkflow` to add rules for a new workflow or update rules for existing workflow.
4. Update `AddWorkflow` to add the workflow rules only if the workflow is new.

AshishPrasad commented:
@abbasc52 - Please review when available.

reachvgupta commented:
Thanks for having this

abbasc52 reviewed (commented):
It would be great if we could add a test case for this scenario to ensure this bug does not reappear in future😄

AshishPrasad commented on src/RulesEngine/RulesEngine.cs:
```diff
@@ -230,7 +230,7 @@ private List<RuleResultTree> ValidateWorkflowAndExecuteRule(string workflowName,
         private bool RegisterRule(string workflowName, params RuleParameter[] ruleParams)
         {
             var compileRulesKey = GetCompiledRulesKey(workflowName, ruleParams);
-            if (_rulesCache.ContainsCompiledRules(compileRulesKey))
```
> could you create a AddOrUpdateWorkflow method in RulesEngine class to expose this functionality?
Added in commit #2.

>I will fix the AddWorkflow method to strictly add only if the workflow is new in a separate PR.
Added in commit #2.

AshishPrasad commented on src/RulesEngine/RulesCache.cs:
```diff
@@ -46,15 +46,35 @@ public bool ContainsCompiledRules(string workflowName)
+        /// <param name="workflowName">The workflow name.</param>
+         /// <returns>
+        ///   <c>true</c> if [compiled rules] is newer than the [workflow rules]; otherwise, <c>false</c>.</returns>
+        public bool IsCompiledRulesUpToDate(string compiledRuleKey, string workflowName)
```
Resolved in commit 2.

AshishPrasad commented:
> It would be great if we could add a test case for this scenario to ensure this bug does not reappear in future😄

Added in commit 2.

AshishPrasad commented:
@abbasc52 - Thank you for you suggestions. Incorporated them in commit 2. Also, updated `AddWorkflow` to only add workflows if the workflow name is new.

Please review when available.

abbasc52 commented on src/RulesEngine/RulesCache.cs:
```diff
@@ -102,19 +125,19 @@ public WorkflowRules GetWorkFlowRules(string workflowName)
         public IDictionary<string, RuleFunc<RuleResultTree>> GetCompiledRules(string compiledRulesKey)
         {
-            return _compileRules[compiledRulesKey];
+            return _compileRules[compiledRulesKey].Item1;
```
Can you add a null check here? It might throw error when compiled rule does not exist

abbasc52 commented on src/RulesEngine/RulesEngine.cs:
```diff
@@ -149,11 +149,37 @@ private async ValueTask<ActionRuleResult> ExecuteActionForRuleResult(RuleResultT
+            {
+                foreach (var workflowRule in workflowRules)
+                {
+                    if (!_rulesCache.ContainsWorkflowRules(workflowRule.WorkflowName))
```
I think we should validate the workflow first before checking it exists.
WorkflowName can be null if input is incorrect

abbasc52 commented:
@AshishPrasad this looks good. Thanks for your efforts! Shall we go ahead and merge it?

AshishPrasad commented:
> @AshishPrasad this looks good. Thanks for your efforts! Shall we go ahead and merge it?

Yes @abbasc52 - please merge the changes and release a new version in order to execute refreshed rules via the new `AddOrUpdateWorkflow` without the need for any synchronization constructs at the call site.
